File: activity scheduler/activityScheduler.cpp
/******************************************************
Program Name: activity scheduler
Author: Christian Roccanova
Date: 4/28/2018
Description: Reads data from file into an array of and then sorts via insertion sort based on latest starting time.  
				Sorted data is then sent to a greedy algorithm to schedule the optimal amount of activities.
******************************************************/

#include <iostream>
#include <string>
#include <fstream>

using std::cin;
using std::cout;
using std::endl;
using std::string;
using std::ofstream;
using std::ifstream;

//activity object containing its number, start time and finish time
class activity
{
public:

	int num;
	int start;
	int end;
};

//sorts activities by start time with the latest time being first
void insertSort(activity arr[], int len)
{
	activity temp;
	for (int i = 0; i < len; i++)
	{
		for (int j = i; j > 0; j--)
		{
			//swaps positions if a number is larger than the one preceeding it
			if (arr[j].start > arr[j - 1].start)
			{
				temp = arr[j];
				arr[j] = arr[j - 1];
				arr[j - 1] = temp;
			}
		}
	}
}

//selects activities by latest start time
void activitySelect(activity act[], int len)
{
	int count = 0;	
	activity order[10000];

	//activity with the latest start time will always be first
	order[0] = act[0];

	for (int i = 1; i < len; i++)
	{
		//adds activity to the final list if it's end time is compatible with (earlier than or equal to) the previous activitie's start time
		if (act[i].end <= order[count].start)
		{
			count++;
			order[count] = act[i];
			
		}
	}

	//print results
	cout << "Number of activities selected: " << count + 1 << endl;
	cout << "Order of activities: ";
	for (int i = count; i >= 0; i--)
	{
		cout << order[i].num << " ";
	}

	cout << endl << endl;
}



int main()
{
	int actNum;
	activity actArray[10000];
	ifstream dataFile;
	
	dataFile.open("act.txt");
	while (!dataFile.eof())
	{
		dataFile >> actNum;
		
		// fills object array with data
		for (int i = 0; i < actNum; i++)
		{
			dataFile >> actArray[i].num;
			dataFile >> actArray[i].start;
			dataFile >> actArray[i].end;
								
		}

		//sorts array
		insertSort(actArray, actNum);

		//schedules activities
		activitySelect(actArray, actNum);

	}
	
	
	return 0;
}

File: activity scheduler/activityScheduler.py
# Christian Roccanova
# CS325 - fall 2018
# last-to-start - reads a set of activities with start and end times from a file.  These are then sorted via merge sort and a greedy algorithm selects the gretest possible number of activities based on latest start time.


#activity class definition
class activity:
        def __init__(self, number, start, finish):
            self.num = number
            self.start = start
            self.fin = finish

#mergeSort code taken from HW 1
#merges left and right sub arrays
def merge(data, left, right):
     
       #array indices
       i = 0
       j = 0
       k = 0

       sizeL = len(left)
       sizeR = len(right)

       #merge sub arrays 
       while i < sizeL and j < sizeR:
           if left[i].start > right[j].start:   #flipped < to > to make descending order based on start time
               data[k] = left[i]
               i += 1
           else:
               data[k] = right[j]
               j += 1
           k += 1

       #copy remaining elements of left sub array
       while i < sizeL:
           data[k] = left[i]
           i += 1
           k += 1
       #copy remaining elements of right sub array
       while j < sizeR:
           data[k] = right[j]
           j += 1
           k += 1

#recursively sorts array by splitting it into sub arrays, sorting those and then re-combining them
def mergeSort(data):

   dataSize = len(data)

   #if only one element, its already sorted
   if dataSize > 1:
       
       #split passed data into left and right halves
       mid = int(dataSize/2) #get position of middle and force it to be an int
       left = data[:mid] #array of all elements from position 0 to mid
       right = data[mid:] #array of all elements from mid to last element


       #recursive calls on each split
       mergeSort(left)
       mergeSort(right)

       #re-combines sub arrays
       merge(data, left, right)


#greedy algorithm to select the gretest possible number of activities based on latest start time
def selection(activities):

    #sort in descending order by start time
    mergeSort(activities)

    #initialize new schedule
    schedule = []
    key = 0

    #first activity is always chosen
    schedule.append(activities[0])
    selectCount = 1

    #loop through remaining activities, adding them to the schedule if their end time is at or before the start time of the most recent addition
    for i in range(1, len(activities)):
        if activities[key].start >= activities[i].fin:
            schedule.append(activities[i])
            key = i
            selectCount += 1

    #flip schedule so that activities are in order from start to finish
    schedule.reverse()

    #print data to terminal
    print("Number of activities selected = ", selectCount )
    print("Activities selected:", end='')
    for x in range(0, selectCount):        
        print(" ", schedule[x].num, end='')

    print("\n")


# Main
# get data from file
dataFile = open("act.txt", "r")
lines = dataFile.readlines()
dataFile.close()
lineCount = len(lines)


j = 0
setCount = 1
while j < lineCount:

    activities = []    
    actCount = int(lines[j])

    #build activity list
    for k in range(j + 1, j + actCount + 1):
        currLine = lines[k].split()
        actNum = int(currLine[0])
        actStart = int(currLine[1])
        actFin = int(currLine[2])

        activities.append(activity(actNum, actStart, actFin))

    j = j + actCount + 1

    print("Set ", setCount)

    #call selection function
    selection(activities)
    setCount += 1

